canonical-web-and-design/snapcraft.io
# [WIP] Set phased release on promotion, show phased release in table

Fixes https://github.com/canonical-web-and-design/snapcraft.io/issues/2239
Fixes https://github.com/canonical-web-and-design/snapcraft.io/issues/2240

## QA
- Pull the branch
- `./run`
- Visit http://0.0.0.0:8004/snap/releases
- Drag a revision to a different channel
- The confirmation bar should show some options
- Set the % (or don't) and see if what happens when you release is expected
- Do the same for multiple revisions
- Generally use the feature and see what happens

## TO NOTE
- You should only get the option to phase if there is already a release in a channel & arch
  - On this note, if rev 3 is on latest/edge and there is no rev on latest/beta and you drag rev 3 onto both latest/edge and latest/beta the phasing dialog should display "Release 1 revision to..."
  - IF however you have rev 2 in latest/beta the dialog should display "Release 2 revisions to..." and the tooltip should list each channel separately.
- You should not be able to drag any single revision onto a channel & arch that is being phased
- You should not be able to drag an entire channel (or use the context menu) to a channel that has a phased release in progress - the context menu should give you the reason why.

## TODO
- [x] Prevent showing phasing options for the following cases:
  - [x] The channel has never had anything released before or is closed
  - [x] You're promoting a revision to a closed channel that is already tracking the same revision in a lower risk
- [ ] Phasing can never truly be 100%, this is a hard concept to explain succinctly in the UI - should we display "99.99%"?
- [x] Preventing dragging of revisions to channel/arch's that are already being phased
